### PR TITLE
Reader move tab layout to parent part 6 - search

### LINKED ISSUE #11849: Reader Master Issue: Enable horizontal swipe gesture in the Reader
Goal of this issue is to enable swipe gesture on the Reader and prepare the ground for the Reader Improvements project. We plan to refactor/rewrite one tab (following, discover, ..) at a time so we can release the changes continuously.
The plan for this PR is to introduce a new parent fragment with a tablayout - each tab would be an instance of ReaderPostListFragment. We'll need to do several changes in order to support having multiple instances of the ReaderPostListFragment at the same time. The current implementation uses a single instance of the ReaderPostListFragment and only changes filtering criteria when a different tab is selected.

Note: We started tracking the related PRs and issues in this ticket. However, as the number of issues increased we decided to start creating real issues and adding them to the "Reader Improvements" board. So the list below is not complete.

Created PRs
- https://github.com/wordpress-mobile/WordPress-Android/pull/11840 - Introduce Tablayout in parent fragment
- https://github.com/wordpress-mobile/WordPress-Android/pull/11848 - Remove unnecessary code from ReaderPostListFragment
- https://github.com/wordpress-mobile/WordPress-Android/pull/11867 - Move NewsCard to ReaderFragment
- https://github.com/wordpress-mobile/WordPress-Android/pull/11881 - Bug fixes
- https://github.com/wordpress-mobile/WordPress-Android/pull/11890 - Move SubFilter component into it's own VM
 - https://github.com/wordpress-mobile/WordPress-Android/pull/11890 - Restore lastly selected tab
- https://github.com/wordpress-mobile/WordPress-Android/pull/11902#pullrequestreview-410802534 - Fix "discover" button on the empty screen

Remaining issues
- [x] SubFilter component on the following tab doesn't show any data (the issue is that ReaderPostListViewModel is tied to ReaderPostListFragment's scope but the component is looking for it in WPMainActivity) (fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11872)
- [x] Move "NewsCard" to the ReaderFragment - currently it's shown on all the tabs (fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11867)
- [x] Verify a correct tag is loaded in the subfilter component when you add a new tag in the tag settings(fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11872)
- [x] Move TabLayout into the ReaderFragment (fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11840)
- [x] Move Toolbar into the ReaderFragment (fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11899)
- [x] Revert b558f2a388932508dcb966c20578cf2772be21b2 this commit and come up with a proper fix. (Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11881)
- [x] If I keep switching tabs long enough the app sometime shows "discover" content on all the tabs (it keeps happening quite often on self-hosted site) (fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11890)
- [x] Store and reopen lastly selected tab after app restart (fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11890)
- [x] Verify dark mode - especially subfilter component (fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11899)
---------------------------------------
- [-] The tabs names are different from before (not sure if it is intended), I get the below instead of the Following, Discover, Likes, Saved names (check what field we use for the tab names - also check if translations work, because I had troubles with translations when using a different field for the names) (created issue - https://github.com/wordpress-mobile/WordPress-Android/issues/11905)
- [-] "Time spent in reader" tracking is broken now, since we have multiple instances of the fragment. It should be moved to the ReaderFragment. (created issue - https://github.com/wordpress-mobile/WordPress-Android/issues/11906)
- [-] Saved posts empty screen - "discover" button doesn't work (created issue - https://github.com/wordpress-mobile/WordPress-Android/issues/11900)
- [-] When the user restarts the app and half-swipes in the reader the next fragment isn't pre-loaded (created issue - https://github.com/wordpress-mobile/WordPress-Android/issues/11907)
- [-] Move the Search into the ReaderFragment (Created ticket https://github.com/wordpress-mobile/WordPress-Android/issues/11908)
- [-] Initial/default tab should be discover (created ticket - https://github.com/wordpress-mobile/WordPress-Android/issues/11909)
- [-] When you sign up with a new user and you have no sites the empty states and the content can overlap like this below (they are drawn on top of each other) - HIGH PRIORITY! (created issue - https://github.com/wordpress-mobile/WordPress-Android/issues/11910)
- [-] Verify deeplinks and regular `startActivity` navigation works as expected (created notes on the ReaderImprovements board)
- [-] Verify re-blog feature works as expected(created notes on the ReaderImprovements board)

### LINKED ISSUE #11908: Reader: Move search from the ReaderPostListFragment into ReaderFragment
The "search" in the Reader is still located in the ReaderPostListFragment but isn't visible on the screen anymore (due to the refactoring). Goal of this task is to introduce search in the ReaderFragment - either by moving it or writing it from scratch.

### PR DESCRIPTION
**Pull request by malinajirka**
Fixes #11908
Partially fixes #11849

**Merge instructions:**
1. Review this PR
1. ~Make sure https://github.com/wordpress-mobile/WordPress-Android/pull/11899 is merged~
2. ~Change the target branch to master-feature-branch/move-reader-tablayout-to-parent~
3. ~Remove the "Not Ready for merge" label~
4. Merge this PR

This is another PR in a series of PRs which is trying to introduce a regular TabLayout+ViewPager2 in the main Reader view. The current implementation uses FilteredRecyclerView component that has some downsides (eg. swipe gesture doesn't work). These changes will also help us implement the improvements in the ReaderImprovements project.
 
P.S. The known issues/remaining tasks are listed in the ReaderImprovements project board.

Goal of this PR is to re-introduce the search feature in the Reader. I was planning to extract the feature from the ReaderPostListFragment or even rewrite it from scratch. As it turned out this task would be much bigger than I anticipated. The feature is quite complex and contains a lot of workarounds for edge cases. I decided to reuse the existing functionality and just re-wire the navigation.
This solution has one minor downside I'm aware of -> when the user clicks on the search icon a new screen is displayed and the default SearchView animations isn't played. This could be potentially solved by using `fragment transition animations` but I decided against it to keep the code as simple as possible. The fragment/activity transition looks good enough imo. 

To test:
1. Open Reader
2. Click on Search
3. Type something
4. Click on the clear button
5. Type something
6. Click on Done
7. Notice the search is performed
8. Scroll and switch tabs
9. Rotate your device and make sure the scroll position is retained and the app doesn't crash :P
10. Restart the app
11. Open Reader
12. Click on Search
13. Notice the previously searched query is being suggested
14. Type a first letter of the same query
15. Notice auto-completion works

Continue testing whatever you can think of.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java [hunk 1]
```diff
@@ -423,22 +434,19 @@ public void onCreate(Bundle savedInstanceState) {
...
 
-        mViewModel.start(!mIsTopLevel, mReaderViewModel);
+        mViewModel.start(mReaderViewModel);
 
         if (isFollowingScreen()) {
             mSubFilterViewModel.onUserComesToReader();
         }
+
+        if (getPostListType() == ReaderPostListType.SEARCH_RESULTS) {
+            mRecyclerView.showAppBarLayout();
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:446]: We want to start the screen with the SearchView expanded.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java [hunk 2]
```diff
@@ -1124,35 +1128,8 @@ public boolean onMenuItemActionExpand(MenuItem item) {
...
-                    }
-
-                    mViewModel.onSearchMenuCollapse(true);
-                } else {
-                    // return to the followed tag that was showing prior to searching
-                    resetPostAdapter(ReaderPostListType.TAG_FOLLOWED);
-                }
-
-                return true;
+                requireActivity().finish();
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:1131]: When the user collapses the SearchView, we want to finish this activity and go back to WPMainActivity + ReaderFragment -> we don't need to do any cleanup as the fragment dies.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java [hunk 3]
```diff
@@ -423,22 +434,19 @@ public void onCreate(Bundle savedInstanceState) {
...
+        mViewModel.start(mReaderViewModel);
 
         if (isFollowingScreen()) {
             mSubFilterViewModel.onUserComesToReader();
         }
+
+        if (getPostListType() == ReaderPostListType.SEARCH_RESULTS) {
+            mRecyclerView.showAppBarLayout();
+            mSearchMenuItem.expandActionView();
+            mRecyclerView.setToolbarScrollFlags(0);
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:448]: The Toolbar shouldn't scroll with the content of the search suggestions and search results -> the search input is in the toolbar and it should be always visible so the user can easily get back to it (this is the same behavior as in develop).

**Comment by zwarm** [comment]
@malinajirka - I'm not seeing the search icon in the attached APK

Steps to reproduce
- Fresh login
- Tap Reader 
- Notice that the search icon does not appear
- Background app and then close (by swiping away)
- Reopen app - Search icon shows in the reader

**Comment by malinajirka** [comment]
I haven't noticed this while testing the app. It seems the icon doesn't show up only in certain scenarios. Either case, this issue was caused by a bug in the BottomNavigationComponent. I've merged `develop` into this branch and all should be good now. Let me know if you encounter this issue again. Thanks @zwarm ;)

**Comment by zwarm** [comment]
@malinajirka - Seeing some weird behavior when rotating the app while viewing search results: the `sites` tab doesn't maintain its search data

**Steps to reproduce - Emulator API 29, Samsung S20 & Google Pixel 3XL**
- Tap search icon
- Enter search word (I used bacon)
- Notice that results load for the Posts tab
- Tap the Sites tab
- Rotate the device
- Notice that the Sites list is now empty and/ or shows a message "No results found for bacon for your language"

**Comment by malinajirka** [comment]
Thanks @zwarm, I've encountered this issue as well. I created a bug yesterday and I assigned it to Groundskeeping, since the issue is happening even on `develop` - https://github.com/wordpress-mobile/WordPress-Android/issues/11913.

**Comment by zwarm** [comment]
> I created a bug yesterday  
👍 
Then everything else is ready to go. @malinajirka

**Review by zwarm — APPROVED** [review]
LGTM!

**Pull request merged**